### ISSUE TITLE
dbplyr summarize() without group_by()

### ISSUE THREAD

**Issue by llendway** [beanumber/mdsr#62]
Hi all! I posted a question over at [RStudio Community](https://community.rstudio.com/t/dbplyr-summarize-without-a-group-by/116466) but it turns out it belongs here. I've also included the same reprex from that post below. 

library(tidyverse)   # for reading in data, graphing, and cleaning
library(dbplyr)      # for SQL query "cheating" - part of tidyverse but needs to be loaded separately
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql
library(mdsr)        # for accessing some databases - goes with Modern Data Science with R textbook
library(RMySQL)      # for accessing MySQL databases
#> Loading required package: DBI
library(RSQLite)     # for accessing SQLite databases
#> 
#> Attaching package: 'RSQLite'
#> The following object is masked from 'package:RMySQL':
#> 
#>     isIdCurrent

con_air <- dbConnect_scidb("airlines")

overall_smry <-
  tbl(con_air, "flights") %>% 
  summarize(min_year = min(year),
            max_year = max(year),
            num_flights = n(),
            num_carriers = n_distinct(carrier))

overall_smry
#> Warning: Missing values are always removed in SQL.
#> Use `MIN(x, na.rm = TRUE)` to silence this warning
#> This warning is displayed only once per session.
#> Warning: Missing values are always removed in SQL.
#> Use `MAX(x, na.rm = TRUE)` to silence this warning
#> This warning is displayed only once per session.
#> Error in .local(conn, statement, ...): could not run statement: Unknown column '' in 'field list'

overall_smry %>% 
  show_query()
#> <SQL>
#> SELECT ``, MIN(`year`) AS `min_year`, MAX(`year`) AS `max_year`, COUNT(*) AS `num_flights`, COUNT(DISTINCT `carrier`) AS `num_carriers`
#> FROM `flights`
#> GROUP BY ``
Created on 2021-09-28 by the reprex package (v2.0.0)

**Comment by nicholasjhorton** [comment]
Lisa, sorry for the complications.

Ben looked at this a few months ago:


> On Jul 10, 2021, at 7:22 AM, Ben Baumer ***@***.***> wrote:
> 
> Huh. If you add "show_query()" to the end of your pipeline, you can see what is happening. 
> 
> > PI_Info %>% summarize(total = n()) %>% show_query()
> <SQL>
> SELECT ``, COUNT(*) AS `total`
> FROM `PI_Info`
> GROUP BY ``
> 
> This produces an invalid SQL query. 
> 
> This hacky workaround will get you the answer you want:
> 
> > PI_Info %>% group_by("a") %>% summarize(total = n())
> # Source:   lazy query [?? x 2]
> # Database: mysql 5.5.58-0ubuntu0.14.04.1-log ***@***.***:/wai]
>   `"a"` total
>   <chr> <dbl>
> 1 a        20
> 
> I'm not sure why it wants to insert a GROUP BY when the data.frame that is being piped into summarize() isn't grouped? 
> 
> -Ben

Does that help?

I'm still puzzled about whether this is a dplyr/dbplyr issue or an issue with the two MySQL databases I routinely use.

**Comment by llendway** [comment]
Yes. And, like you see in that RStudio community post, it looks like it's a MySQL issue. For my class, I'm just going to skip that one query, which really isn't a big deal. If need be, I'll use the "hacky" solution - what's a data science course without at least a few of those!

**Comment by nicholasjhorton** [comment]
But it looks like there's a simple workaround:

```
dcon <- function(dbname, ...){
  {
    DBI::dbConnect(RMariaDB::MariaDB(), dbname = dbname, host = "mdsr.cdc7tgkkqd0n.us-east-1.rds.amazonaws.com", 
                   user = "mdsr_public", password = "ImhsmflMDSwR", 
                   ...)
  }
}
```

this can be used instead of con_air <- dbConnect_scidb("airlines")
i.e. after the above definition do

```
con_air <- dcon("airlines")
```

**Comment by nicholasjhorton** [comment]
Here's an updated reprex which demonstrates the fix. 

@beanumber might you be willing to incorporate this change into the mdsr package?

``` r
library(tidyverse) # for reading in data, graphing, and cleaning
suppressPackageStartupMessages(library(dbplyr)) # for SQL query "cheating" - part of tidyverse but needs to be loaded separately
library(mdsr) # for accessing some databases - goes with Modern Data Science with R textbook
library(RMySQL) # for accessing MySQL databases
#> Loading required package: DBI
suppressPackageStartupMessages(library(RSQLite)) # for accessing SQLite databases

dcon <- function(dbname, ...){
  {
    DBI::dbConnect(RMariaDB::MariaDB(), dbname = dbname, host = "mdsr.cdc7tgkkqd0n.us-east-1.rds.amazonaws.com", 
                   user = "mdsr_public", password = "ImhsmflMDSwR", 
                   ...)
  }
}

con_air_mdsr <- dbConnect_scidb("airlines")
con_air <- dcon("airlines")

tbl(con_air_mdsr, "flights") %>%  # doesn't work
  summarize(min_year = min(year, na.rm = TRUE),
            max_year = max(year, na.rm = TRUE),
            num_flights = n(),
            num_carriers = n_distinct(carrier))
#> Error in .local(conn, statement, ...): could not run statement: Unknown column '' in 'field list'
tbl(con_air, "flights") %>%  # works
  summarize(min_year = min(year, na.rm = TRUE),
            max_year = max(year, na.rm = TRUE),
            num_flights = n(),
            num_carriers = n_distinct(carrier))
#> # Source:   lazy query [?? x 4]
#> # Database: mysql
#> #   [mdsr_public@mdsr.cdc7tgkkqd0n.us-east-1.rds.amazonaws.com:NA/airlines]
#>   min_year max_year num_flights num_carriers
#>      <int>    <int>     <int64>      <int64>
#> 1     2010     2017    47932811           20
```

<sup>Created on 2021-09-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by beanumber** [comment]
Sure, I can do that.

**Comment by nicholasjhorton** [comment]
I know that you are swamped, but wonder if there's ways to move this forward at some juncture.

**Comment by kcarnold** [comment]
Yup, just using `RMariaDB::MariaDB()` instead of RMySQL works. I ran an exercise in class yesterday with this solution and everything worked great. (Thanks for hosting this database!)

**Comment by beanumber** [comment]
Thanks @nicholasjhorton @kcarnold @llendway. 

I pushed a fix on the `maria` branch: 831b58c

I'll merge once the semester is safely over.

**Comment by llendway** [comment]
Thank you! And, yes, never modify anything until the semester is over - good rule to live by :)

**Comment by beanumber** [comment]
Closed via aea931b
